### ISSUE TITLE
Bug: [await-thenable] inconsistent behavior on promise unions

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#11609]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=MYewdgzgLgBAtgQwJ4CMCmAFATiOBLCNGAXhgFkEoALAOiwTABNcAKAShgH4ZtcC06aCCAA2ANzTsYALhgBGANwAoJQgDuCPLESpMOfIQUwA9MZhYAriKLMhMMCFhY0ABxBYoK3gYEIRIlgBtHXRvfgBdNiNTcysiZzcPCBU0AA9E2ABvAF8FIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOiiJo0APbRI4MAF8QSoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const maybePromise = Math.random() ? Promise.resolve() : 1;

await maybePromise; // rule does not report

Promise.all([maybePromise]); // rule reports

export {};
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/await-thenable": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The rule should not report.

### Actual Result

The rule reports on passing `maybePromise` into `Promise.all`.

### Additional Info

In #11267, @ronami added support for non-promise values passed to aggregator methods.

However, this now introduces an inconsistency:

- The rule correctly does _not_ report when `await`ing something that might be a `Promise`.
- However, the rule wrongly reports when passing something that might be a `Promise` into a promise aggregator function.

**Comment by branko-d** [comment]
After upgrading typescript-eslint from 8.41.0 to 8.44.0, we started getting the same error on the code that previously passed.

This code constructs an array where each element is either `Promise` or `undefined`, in a way that cannot be predicted in advance. We then `Promise.all` on the array and get:

    132:47  error  Unexpected iterable of non-Promise (non-"Thenable") values passed to promise aggregator  @typescript-eslint/await-thenable

The workaround is to convert `undefined` elements to promises (via `Promise.resolve(undefined)`), but that just creates extra objects for no reason. I think the `await-thenable` rule should be changed to pass when an array element _can_ be promise (instead of passing only when it _must_ be promise).

**Comment by JoshuaKGoldberg** [comment]
cc @ronami / #11267, wdyt?

**Comment by kirkwaiblinger** [comment]
Marked accepting PRs. I think during development one of the motivations for being strict like this was because doing something like
```ts
Promise.all([3, Promise.resolve(4)]);
```
looks roughly like an `Array<number | Promise<number>>`, and it makes sense to report the `3`. But we ended up special-casing array literals to check each element individually, so we can relax it to allowing a union now without failing to report on the above.

**Comment by ronami** [comment]
> cc [@ronami](https://github.com/ronami) / [#11267](https://github.com/typescript-eslint/typescript-eslint/pull/11267), wdyt?

Hmm I think there's some trade-off between safety and convenience, as allowing values which may not be a promise can be a bug/mistake.

However, I agree with @kirkwaiblinger that this would otherwise make the rule pretty impractical to use 👍

**Comment by gustavopch** [comment]
Still happens to me in this situation:
- typescript v5.9.2
- typescript-eslint v8.44.1
```ts
declare const getItem: (id: string) => Promise<unknown>

const results = await Promise.all(
  // Unexpected iterable of non-Promise (non-"Thenable") values passed to promise aggregator. eslint(@typescript-eslint/await-thenable)
  ['1', '2'].map(id => (id ? getItem(id) : null)),
)
```

**Comment by valadaptive** [comment]
This should be reopened. #11611 only fixes it for array *literals*--a non-literal array type whose elements may or may not be promises will still error. See [this line](https://github.com/valadaptive/sb3-js/blob/8bacb4c6d57a0da02dac40fdf7d1544eb5f52bb1/src/load-svg.ts#L50), for instance.

**Comment by valadaptive** [comment]
I also have to question the decision to make the `Promise.all` behavior 1) part of the `await-thenable` rule and 2) not configurable at all. The `await-thenable` rule is explicitly about the `await` keyword, and [the documentation](https://typescript-eslint.io/rules/await-thenable/) says so and doesn't even mention promise aggregator methods anywhere.

Somehow, there's now this extra lint rule that doesn't even require the `await` keyword, rejects useful and meaningful code, and is not configurable, and I cannot figure out why it was made part of the `await-thenable` rule.

**Comment by kirkwaiblinger** [comment]
@gustavopch or @valadaptive 

Would you kindly open new issue(s) to discuss further? 

See also https://typescript-eslint.io/contributing/issues

Thanks!

**Comment by gustavopch** [comment]
@kirkwaiblinger I'm not sure to what part of the linked page you're referring. I see "Search all *opened and closed* issues to make sure your issue wouldn't be a duplicate" there and when I try to open a "Rule Bug" issue, there's a checkbox where I have to confirm "I have searched for related issues and found none that matched my issue". The snippet I sent above is a repro for the exact problem described in this issue, not a different problem (AFAICT at least), so I'm a bit confused. Anyway, it seems @valadaptive already opened another issue, but I thought I should leave this as feedback.
